### ISSUE TITLE
PDF.js does not take into account /ActualText

### ISSUE THREAD

**Issue by TimotheAlbouy** [mozilla/pdf.js#12237]
[ESET_Okrum_and_Ketrican.pdf](https://github.com/mozilla/pdf.js/files/5089612/ESET_Okrum_and_Ketrican.pdf)

I'm using PDF.js 2.4.456. The issue I'm describing here is independent of the OS and the web browser.

To reproduce the problem, open the attached document with PDF.js (a quick way to reproduce the bug is to open the document in Firefox), then copy and paste some of the text. In the extracted text, the period characters are missing for the main font:

> The Ke3chang group, also known as APT15, is a threat group believed to be operating out of China  Its attacks were first reported in 2012,

This is due to a bad mapping in the `/ToUnicode` property of the font: the `0x2E` code (i.e. the period character in ASCII) in the text is mapped to `U+0020` (a space in UTF-16).

But when we copy-paste the content of the PDF using Adobe Acrobat, we extract all the period characters correctly. It is because Acrobat takes into account the `/ActualText` marked text properties inside the PDF.

This is what we see when we open the document using a PDF inspector like PDFDebugger of PDFBox:

```
(, also known as APT15, is a threat group believed to be operating out of China)Tj
/Span<</ActualText<FEFF002E>>> BDC 
(.)Tj
EMC
```

We see that the 0x2E code in `(.)Tj`, which according to the `/ToUnicode` map represents a space character, is marked to *actually* represent U+FEFF (a BOM) and U+002E in, a period character.

Thus, Acrobat Reader extract the periods correctly in the given report because it takes into account the `/ActualText` content, whereas PDF.js doesn't extract the periods because it only considers the `/ToUnicode` map.

Is there an ongoing effort to fix this problem?

**Comment by Snuffleupagus** [comment]
Duplicate of #12100 (please don't knowingly open duplicate issues).

**Comment by TimotheAlbouy** [comment]
It wasn't clear that the issue came from the ActualText property before, we thought that PDF.js did the job right because it followed the ToUnicode map. I think I would rather close/delete the old issue and leave this one which is clearer, if you agree.

**Comment by timvandermeij** [comment]
I closed the other issue since this one contains a bit more details, but we should indeed keep the discussion limited to this issue from now on.

**Comment by Mowmowj** [comment]
would it be repair in future? @Snuffleupagus  
do you have any idea about it? @TimotheAlbouy

**Comment by Mowmowj** [comment]
hi @TimotheAlbouy  would you think it's possible to do a preliminary work to fix the unicode mapping problem of the pdf file itself? then may the pdf.js would render the text layer normally.

**Comment by TimotheAlbouy** [comment]
@Mowmowj of course this would be possible, but it would not address the root problem: currently PDF.js cannot extract some text that Acrobat Reader can. You wouldn't tell a visually impaired person to change their book instead of replacing their faulty glasses.

**Comment by joelostblom** [comment]
I ran into this issue when I noticed that PDF.js can't find text strings such as `fi`, `tt`, etc if these are represented as ligatures in the PDF documents. I noticed that the PDF viewer Evince is able to find search text even in PDFs that do contain ligatures. Since evince is open source, maybe there is something in their approach that could be useful for PDF.js as well? I'm not familiar with exactly what that would be, but a brief search in their repo shows that they e.g. do some normalization of the text in the PDF https://gitlab.gnome.org/GNOME/evince/-/commit/9de1152cd935d9f00f2709052d25d42b18cb1b0f

**Comment by Mowmowj** [comment]
> I ran into this issue when I noticed that PDF.js can't find text strings such as `fi`, `tt`, etc if these are represented as ligatures in the PDF documents. I noticed that the PDF viewer Evince is able to find search text even in PDFs that do contain ligatures. Since evince is open source, maybe there is something in their approach that could be useful for PDF.js as well? I'm not familiar with exactly what that would be, but a brief search in their repo shows that they e.g. do some normalization of the text in the PDF https://gitlab.gnome.org/GNOME/evince/-/commit/9de1152cd935d9f00f2709052d25d42b18cb1b0f

Good point, looks like it's use the Glib, BTW for inspiration, the mupdf viewer also supports the special string. but i'm not sure how they figure out it.

**Comment by MurakamiShinyu** [comment]
The issue of /ActualText being ignored is also important for CJK language users. There is a related issue in Chromium where /ActualText with CJK characters was ignored, that has recently been fixed. See the Chromium issue:

- [Chrome PDF printing changes CJK characters to Kangxi radicals](https://issues.chromium.org/issues/420508260)

As described in that issue, PDFs generated by Chromium have a problem where some CJK characters are replaced with Kangxi radicals if the PDF viewer does not handle ActualText properly.

The /ActualText in PDF is an essential feature for accessibility, and I hope PDF.js can handle it as well.

**Comment by bungeman** [comment]
To get some idea of the issues one might run into while implementing this I created a patch https://github.com/bungeman/pdf.js/commit/6a5278336d0c1c41b3ad6d264d0c1f29e92e2f4f . This simply reads and stores the `/ActualText` and then adds it as an attribute to the `span` in the text layer. However, it isn't clear exactly how emit the `/ActualText` into the text layer.

**Comment by olyapop** [comment]
Another case: Inter's case-sensitive forms of `(`, `)` and `-` (glyph IDs 0x544, 0x545, 0x55E) have no Unicode mapping in the embedded font's ToUnicode, cmap or post tables, so getTextContent returns U+0000 for them. The PDF wraps each glyph in an /ActualText span with the intended character, and other readers extract it correctly. I reproduced this on 6.3.289. I can't share the file, but I can put together a minimal repro if that would help.
